### PR TITLE
Add device `TargetToken` auth

### PR DESCRIPTION
**Pull request by b-rowan**
Adds support for the `TargetToken` auth header.

This can be set up in 3 modes (or disabled altogether, which is the default):
`setup` - Setup mode is for setting up or migrating from an existing solution (such as hawkbit).  Setup mode will auto register all devices that poll, and attach the auth token they poll with to their device information in the database.
`lax` - Lax mode will handle auth tokens on devices that have them set, and will error if those devices don't send a valid token, but devices that were set up without a token will not require one.
`strict` - Strict mode requires a token for every device.  All devices will encounter errors if they submit invalid tokens, and new devices are not registered via the DDI endpoints.  All devices must be set up via the `/device` PUT endpoint, and all devices must have authentication tokens.

Fixes: #175

### CODE REVIEW

File: goosebit.yaml [hunk 1]
```diff
@@ -1,32 +1,46 @@
 ## Settings to adjust for each installation
+## Settings are listed with single #, comments with 2 #
```
Comment (easybe) [inline goosebit.yaml:2]: All these unrelated changes should go into a separate commit. Anyway, not a fan, sorry.

File: goosebit.yaml [hunk 2]
```diff
@@ -1,32 +1,46 @@
...
 
-# Port to host the server on, default:
+## Port to host the server on, default:
 # port: 60053 # GOOSE ;)
```
Comment (easybe) [inline goosebit.yaml:5]: I suggest this convention as an alternative:
```suggestion
#port: 60053 # GOOSE ;)
```

File: goosebit.yaml [hunk 3]
```diff
@@ -1,32 +1,46 @@
...
+## Whether to track the IP of the device when it polls.  Useful for debugging, but can be turned off for privacy.
 # track_device_ip: true
 
-# Secret key used for parsing user sessions. It is HIGHLY advised to pass this as an environment variable instead.
-# Defaults to a randomized value. If this value is not set, user sessions will not persist when app restarts.
+## Device authentication settings
+# device_auth:
+
+##  Enable the checking of device authentication
+#   enable: false
```
Comment (easybe) [inline goosebit.yaml:27]: And this:
```suggestion
#  Token-based device authentication
#device_auth:
#  enable: false
```

File: goosebit.yaml [hunk 4]
```diff
@@ -1,32 +1,46 @@
...
+
+##  The auth mode, one of:
+##  - strict: Strict mode, all devices must have keys, and all keys must be set via the API.
+##  - lax: Lax mode, devices which have keys must use them, but devices without keys do not require them.
+##  - setup: Setup mode, any devices polling with an auth token that don't have one will save it.
```
Comment (easybe) [inline goosebit.yaml:32]: Is this the same as hawkBit implements it?

Comment (b-rowan) [reply goosebit.yaml:32]: Not actually sure, but my goal is not to be hawkBit, but to be better than hawkBit.  I think this is one of the spots where we have that flexibility to change some things, but this is also why PRs exist, I am open to comment if you guys don't agree.

Comment (easybe) [reply goosebit.yaml:32]: Well, in general, I opt for "as simple as possible" (KISS). IMHO pre-mature optimization should be avoided as much as possible. I would only implement a feature when there is an actual need for it.

I agree, we don't want to be hawkBit. IMHO "simplicity" should be gooseBit's main advantage. However, we should implement certain features (or a subset) in order to facilitate the migration to gooseBit.

Comment (b-rowan) [reply goosebit.yaml:32]: Agreed on the simplicity side. 

The idea was that strict mode is what most people will likely want, so that is the default when unspecified.  Lax mode maybe be a ur least useful mode, but I think the one we would use if we were to use auth mode at all.  Setup mode is basically for migration only, because it makes sure there are no issues when migrate with accidentally sending something wrong to the API.

I'm ok with the only option being strict, but I figured these options are simple and useful enough that I may as well just include them.

**Comment by easybe** [comment]
I think there should be some tests for this, especially because we at Husqvarna won't be using this.

**Comment by b-rowan** [comment]
> I think there should be some tests for this, especially because we at Husqvarna won't be using the feature.

Biggest issue here is that it doesn't seem like we can dynamically update the config.  Might need a separate PR to attach it to the app, then it can be used via `request.app.config` or maybe `request.app.goosebit_conf`

**Comment by b-rowan** [comment]
Tests are broken because it won't let me instantiate a new app for this test file (need a new one to alter the config).  No idea how to fix this.

**Comment by tsagadar** [comment]
Great to have progress here. Unfortunately, could not yet look at the code - should have time on the weekend.

Conceptually I had a few thoughts
- It should be extensible to support certificate authentication as well (will be required by us)
- Cannot see the value that `lax` gives
- Ability to disable authentication is helpful to get something up and running quickly
- Devices that fail to authenticate should be logged as warnings. Alternative is to create them but assign them the state `unauthenticated`. Would simplify spotting problematic gateways but would allow to make a DoS attack by creating massive amounts of devices. A middle ground would be to limit the amount of unauthenticated devices - not sure if this is a good idea.
- `setup` seems convenient. Still not sure if it should be supported though Idea is to use this mode only temporarily. But this means that a mechanism has to be built anyways to pre-populate devices. So why not do it right from the start?

What about the following settings
- `Allow un-authenticated devices (only use in test installations)`
- `Allow device authentication by certificate (through reverse proxy)`
- `Allow device authentication by security token`

As for the behavior: if "allow un-authenticated" is enabled, still fail if reverse-proxy header / token doesn't match.

File: goosebit/ui/bff/devices/requests.py [hunk 5]
```diff
@@ -10,3 +10,4 @@ class DevicesPatchRequest(BaseModel):
     pinned: bool | None = None
     feed: str | None = None
     force_update: bool | None = None
+    auth_token: str | None = None
```
Comment (tsagadar) [inline goosebit/ui/bff/devices/requests.py:13]: Should it be added to the bff even though the UI does not use it? In a way it is dead code like this.

File: goosebit/api/v1/devices/routes.py [hunk 6]
```diff
@@ -48,4 +48,32 @@ async def devices_delete(_: Request, config: DevicesDeleteRequest) -> StatusResp
     return StatusResponse(success=True)
 
 
+@router.patch(
+    "",
+    dependencies=[Security(validate_user_permissions, scopes=["device.write"])],
+)
+async def devices_patch(_: Request, config: DevicesPatchRequest) -> StatusResponse:
```
Comment (tsagadar) [inline goosebit/api/v1/devices/routes.py:55]: Duplicated put endpoint. PATCH IMHO should fail if resource does not exist - which is not the case.

File: goosebit/api/v1/devices/routes.py [hunk 7]
```diff
@@ -48,4 +48,60 @@ async def devices_delete(_: Request, config: DevicesDeleteRequest) -> StatusResp
...
+        if config.feed is not None:
+            await updater.update_feed(config.feed)
+        if config.force_update is not None:
+            await updater.update_force_update(config.force_update)
+        if config.auth_token is not None:
+            await updater.update_auth_token(config.auth_token)
+    return StatusResponse(success=True)
+
+
+@router.put(
```
Comment (tsagadar) [inline goosebit/api/v1/devices/routes.py:79]: Add test case for this?

Comment (tsagadar) [reply goosebit/api/v1/devices/routes.py:79]: Also rather non-standard to have such a "batch" request to update/create multiple resources. But can see the advantages if one wants to modify a big number of devices.

File: goosebit/ui/bff/devices/routes.py [hunk 8]
```diff
@@ -79,6 +79,8 @@ async def devices_patch(_: Request, config: DevicesPatchRequest) -> StatusRespon
             await updater.update_feed(config.feed)
         if config.force_update is not None:
             await updater.update_force_update(config.force_update)
+        if config.auth_token is not None:
```
Comment (tsagadar) [inline goosebit/ui/bff/devices/routes.py:82]: Again: dead code?

File: goosebit/settings/schema.py [hunk 9]
```diff
@@ -24,6 +25,17 @@ def get_json_permissions(self):
         return [str(p) for p in self.permissions]
 
 
+class DeviceAuthMode(StrEnum):
+    SETUP = "setup"  # setup mode, any devices polling with an auth token that don't have one will save it
+    STRICT = "strict"  # all devices must have keys, and all keys must be set up with the API
+    LAX = "lax"  # devices may or may not use keys, but device with keys set must poll with them
```
Comment (tsagadar) [inline goosebit/settings/schema.py:31]: See my comment on the PR - `lax` seems pointless. What about my proposal of a different config structure?

**Review by tsagadar — CHANGES_REQUESTED** [review]

**Comment by b-rowan** [comment]
> Cannot see the value that `lax` gives

Lax mode performs exactly like `Allow un-authenticated` mode as you mention above, its just a lot shorter for having in a config.


> * `setup` seems convenient. Still not sure if it should be supported though Idea is to use this mode only temporarily. But this means that a mechanism has to be built anyways to pre-populate devices. So why not do it right from the start?

Confused what you mean here?  Setup is basically for migration from other services, if you are already using hawkbit and don't want to have to find the UUID and auth token for every device, `setup` is just to make sure there are no errors doing initial device setup.

> * `Allow un-authenticated devices (only use in test installations)`
> 
> * `Allow device authentication by certificate (through reverse proxy)`
> 
> * `Allow device authentication by security token`

This is extremely close to what I have.  Lax is allow unauthenticated, and strict is allow via token.  The only change I would make to allow the suggested configuration is add a sub-config for strict that allows you to configure the proxy headers, which is really only one option, the key used to validate the cert.  This is relatively easy to extend into, and I don't fully understand how you guys are doing it, thus I left it out.


> As for the behavior: if "allow un-authenticated" is enabled, still fail if reverse-proxy header / token doesn't match.

Lax already does this, if the token is set for a device and that token is wrong when the device makes a request, it will fail.

File: goosebit/api/v1/devices/routes.py (hunk 7 above)
Comment (b-rowan) [reply goosebit/api/v1/devices/routes.py:79]: More worried about getting the tests to actually run and not just break when I try to instantiate a separate app, I'm sure something to do with how I'm scoping parts of the tests, but I dont really understand well enough to fix it...

File: goosebit/api/v1/devices/routes.py (hunk 6 above)
Comment (b-rowan) [reply goosebit/api/v1/devices/routes.py:55]: Good call.  Not sure if that's in scope for this PR, can fix it here if you want.

File: goosebit/ui/bff/devices/requests.py (hunk 5 above)
Comment (b-rowan) [reply goosebit/ui/bff/devices/requests.py:13]: Maybe?  AFAIU there may be plans to show it on the UI, but fine to leave out.

File: goosebit/ui/bff/devices/routes.py (hunk 8 above)
Comment (b-rowan) [reply goosebit/ui/bff/devices/routes.py:82]: Maybe?  Should we let the user set the token via UI in the future?  If so this should be here, makes 1 less thing we will have to search for in the future, and the performance hit is negligible.

**Comment by tsagadar** [comment]
> > * `setup` seems convenient. Still not sure if it should be supported though Idea is to use this mode only temporarily. But this means that a mechanism has to be built anyways to pre-populate devices. So why not do it right from the start?
> 
> Confused what you mean here? Setup is basically for migration from other services, if you are already using hawkbit and don't want to have to find the UUID and auth token for every device, `setup` is just to make sure there are no errors doing initial device

It works if one has a fix population in the market and those are all online as well. Then you can use `setup` mode for a few days and switch to strict.

Where it does not work: devices somewhere in a sales channel and can come online at any point in time (over the next few years)

Where it does not safe much effort: device still being manufactured. Would then require to create script to add newly produced device tokens to gooseBit. Having that script implemented would then also allow to load the old devices as well.

**Comment by b-rowan** [comment]
> It works if one has a fix population in the market and those are all online as well. Then you can use `setup` mode for a few days and switch to strict.
> 
> Where it does not work: devices somewhere in a sales channel and can come online at any point in time (over the next few years)
> 
> Where it does not safe much effort: device still being manufactured. Would then require to create script to add newly produced device tokens to gooseBit. Having that script implemented would then also allow to load the old devices as well.

Agreed, but in most cases if you were previously using hawkbit you will have at least a few devices deployed, and will need to write that script anyway.  Setup just gives us a small advantage over other services for initial setup, so unless you think there is a big reason not to include it, I can't see any reason we wouldn't just allow it, right?

File: goosebit/api/v1/devices/routes.py (hunk 7 above)
Comment (tsagadar) [reply goosebit/api/v1/devices/routes.py:79]: Here I used another approach to adjust the settings during the pytest execution: https://github.com/husqvarnagroup/goosebit/pull/new/mm/device_auth

File: goosebit/ui/bff/devices/requests.py (hunk 5 above)
Comment (tsagadar) [reply goosebit/ui/bff/devices/requests.py:13]: Let's add it when we get there. Otherwise: dead code.

Comment (b-rowan) [reply goosebit/ui/bff/devices/requests.py:13]: Used in tests.

File: goosebit/ui/bff/devices/routes.py (hunk 8 above)
Comment (b-rowan) [reply goosebit/ui/bff/devices/routes.py:82]: Used in tests.

File: goosebit/api/v1/devices/routes.py [hunk 10]
```diff
@@ -48,4 +48,60 @@ async def devices_delete(_: Request, config: DevicesDeleteRequest) -> StatusResp
     return StatusResponse(success=True)
 
 
+@router.patch(
```
Comment (tsagadar) [inline goosebit/api/v1/devices/routes.py:51]: This gets added in your first commit 1cb764b1 (message: Add auth token to device database and update migrations). It adds the duplicate for patch and put. Why do we need these two calls? Should not only be the patch code in the commit "Add auth token to API and device patch request" ?

File: goosebit/updater/routes.py [hunk 11]
```diff
@@ -12,14 +12,14 @@
 
 async def log_last_connection(request: Request, dev_id: str):
     device = await get_device(dev_id)
-    if config.track_device_ip:
+    if request.scope["config"].track_device_ip:
```
Comment (tsagadar) [inline goosebit/updater/routes.py:15]: Why do we sometimes import the config to use it and sometime we use the middleware to access it?

**Review by tsagadar — CHANGES_REQUESTED** [review]
Few things to change. To get the changes packaged in the proper commits it can help to make a soft reset to the commit before your changes and recreate the commits one by one (at least in PyCharm it is then quite easy to select which lines should go into which commit)

File: goosebit/updater/routes.py (hunk 11 above)
Comment (b-rowan) [reply goosebit/updater/routes.py:15]: Middleware is slightly more dynamic, and should be used in the routes where the `Request` can be accessed.  Everywhere else can use config because it doesn't need to be dynamic or it doesn't have access to request.

File: goosebit/api/v1/devices/routes.py (hunk 6 above)
Comment (b-rowan) [reply goosebit/api/v1/devices/routes.py:56]: Fixed, patch now raises if the device does not exist.

**Review by tsagadar — APPROVED** [review]
